packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -405,6 +381,83 @@ contract ControllerHelper is FlashControllerHelper, IERC721Receiver {
+        _flashLoan(
+            weth,
+            _params.collateralToFlashloan,
+            uint8(CALLBACK_SOURCE.FLASHLOAN_W_MINT_DEPOSIT_NFT),
+            abi.encode(_params)
+        );
+    }
+
+    function _flashCallback(
+        address _initiator,
+        address, /*_asset*/
+        uint256 _amount,
+        uint256 _premium,
+        uint8 _callSource,
+        bytes memory _calldata
+    ) internal override {
+        if (CALLBACK_SOURCE(_callSource) == CALLBACK_SOURCE.FLASHLOAN_W_MINT_DEPOSIT_NFT) {
+            FlashloanWMintDepositNftParams memory data = abi.decode(_calldata, (FlashloanWMintDepositNftParams));
+
+            // convert flashloaned WETH to ETH
+            IWETH9(weth).withdraw(_amount);
+
+            uint256 vaultId = IController(controller).mintWPowerPerpAmount{value: data.collateralToDeposit}(
+                data.vaultId,
+                data.wPowerPerpAmount,
+                0
+            );
+
+            // LP data.wPowerPerpAmount & data.collateralToLp in Uni v3
+            uint256 amount0Desired = isWethToken0 ? data.collateralToLp : data.wPowerPerpAmount;
+            uint256 amount1Desired = isWethToken0 ? data.wPowerPerpAmount : data.collateralToLp;
+            uint256 uniTokenId = _lpWPowerPerpPool(
+                address(this),
+                data.collateralToLp,
+                amount0Desired,
+                amount1Desired,
+                data.lpAmount0Min,
+                data.lpAmount1Min,
+                block.timestamp,
+                data.lpLowerTick,
+                data.lpUpperTick
+            );
+
+            _checkLpMintExcess(vaultId);
+
+            // deposit Uni NFT token in vault
+            INonfungiblePositionManager(nonfungiblePositionManager).approve(controller, uniTokenId);
+            IController(controller).depositUniPositionToken(vaultId, uniTokenId);
+
+            // remove flashloan amount in ETH from vault + any amount of collateral user want to withdraw (sum <= vault.collateralAmount)
+            IController(controller).withdraw(vaultId, _amount.add(data.collateralToWithdraw));
+
+            // convert flashloaned amount + fee from ETH to WETH to prepare for payback
+            IWETH9(weth).deposit{value: _amount.add(_premium)}();
+
+            // if openeded new vault, transfer vault NFT to user
+            if (data.vaultId == 0) IShortPowerPerp(shortPowerPerp).safeTransferFrom(address(this), _initiator, vaultId);
+        } else if (CALLBACK_SOURCE(_callSource) == CALLBACK_SOURCE.FLASHLOAN_CLOSE_VAULT_LP_NFT) {
+            closeUniLpParams memory data = abi.decode(_calldata, (closeUniLpParams));
+
+            // convert flashloaned WETH to ETH
+            IWETH9(weth).withdraw(_amount);
+
+            IController(controller).deposit{value: _amount}(data.vaultId);
+
+            IController(controller).withdrawUniPositionToken(data.vaultId);
+
+            _closeUniLp(data);
```

Comment:
we have to re-add the uniLP token if it is not a full close (otherwise won't be able to refund flashloan), maybe check if liquidityPercentage is != 1e18 and redeposit if so?

Reply:
don't we also have to withdraw the flashloaned collateral here (and then do the same thing with the weth + premium)?  similar pattern to FLASHLOAN_CLOSE_VAULT_LP_NFT?

Reply:
Can u explain more ur second comment ?

Reply:
we deposit _amount as a flashLoan to have it be collateralized and then we withdraw the uni position token, deal with closing it (fully or partially) and then redeposit it back if not fully close.

we don't ever withdraw _amount of ETH from the vault, so we won't pay back the flashloan right?  We also need to convert the  premium and _amount to repay to weth.

similar to this:

```
            IController(controller).withdraw(vaultId, _amount.add(data.collateralToWithdraw));

            // convert flashloaned amount + fee from ETH to WETH to prepare for payback
            IWETH9(weth).deposit{value: _amount.add(_premium)}();
```